### PR TITLE
[video] Remove unnecessary stream details extraction in version/extra addition

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

* Skip stream details extraction from the video when a file added to a movie as version or extra is already a version or extra of another movie.
The stream details were extracted the first time the file was added as version or extra to a movie.

There are more situation where the extraction could be skipped but a db query would be needed to test. This imperfect test comes free because derived from a query earlier in the procedure (GetVideoVersionInfo()).

~~* second commit: refactor the code adding a file as version or extra. It was kept very similar in the versions and extras manager dialogs to facilitate the merge.~~

~~There are many localized messages and they cause many ugly if (version) else if (extra) conditions but I didn't find a satisfying way to refactor them. About half of the messages are tied to the type of dialog and can be handled decently with class inheritance, but the rest is conditional to the current asset type of the file or follow complex conditions.
Maybe how to handle the messages will become clearer if another type of video assets is ever added.~~

## Motivation and context

Noticed during testing that moving a video asset between movies caused an unnecessary stream details extraction.

~~The code adding a file as movie version or extra could easily get out of sync and create weird behaviors in the future.~~

## How has this been tested?

Exercised the different cases handled by the new AddVideoAssetFilePicker() function (formerly CGUIDialogVideoManagerExtras::AddVideoExtra() and CGUIDialogVideoManagerVersions::AddVideoVersionFilePicker())
new unknow file, file that's a version of the same movie, file that's an extra of the same movie,  file that's a version of a different movie, file that's an extra of a different movie, file that's the default version of another movie.

## What is the effect on users?

Moving a version or extra to a different movie is slightly faster.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
> second commit: refactor the code adding a file as version or extra. It was kept very similar in the versions and extras manager dialogs to facilitate the merge.

> There are many localized messages and they cause many ugly if (version) else if (extra) conditions but I didn't find a satisfying way to refactor them.

I see what you are trying to achieve, but I have to admit that I think the new code is even worse than the old code. It looks not simpler to me and you are violating a fundamental principle of object oriented programming now, by introducing knowledge about derived classes (specializations) in the base class. This is imo a no go.

BTW: You have done this design flaw already with another PR, but I missed that at that point in time.

**Comment by CrystalP** [comment]
The two functions in the derived dialogs are 95% identical, I wanted to solve that duplication problem, I may be aware of it, someone else may not. I can't add a comment every second line that it's the same in the other file.  I know it doesn't look any simpler, the logic itself is lengthy.

There are only a couple messages for which your comment is true (the half I mentioned), and I can change them to take advantage of inheritance (a map of messages numbers filled by the derived classes, looked up by the parent). Yes I also identified that remove fits that mold.

The processing is dialog "asset type"-neutral otherwise. There is now a variable assetType where previously the two inherited dialogs had VideoAssetType::VERSION or VideoAssetType::EXTRA in many places.

I know the function is not great, but it's the cost of the flexibility of the design: any file can be added as version or extra, whether it's in a library, in a source, or in any random directory, What looks complicated and difficult to read are the conditions to warn the user of actions he may not intend or situations he may not be aware of and to block the few that can't be done, and provide a helpful (though not always simple) message instead of a generic "can't do that" message.

Part of the solution would be to place some limits on the files that can be turned into a version or extra and make the scanner grab them and not allow another way. Rules like "extras can only reside in subdirectories of an Extras folder in the movie directory".

**Comment by CrystalP** [comment]
Additional commit respecting the messages "inheritance". Could be done by a function overriden by the derived classes instead of a map. The idea of enum values is necessary I think to avoid one overriden function by message id.

File: xbmc/video/dialogs/GUIDialogVideoManager.h
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManager.h:78]: Could this be an `unordered_map` ?

File: xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp:46]: Move this to the initialzer list, please.

File: xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:49]: Move this to the initialzer list, please.

**Review by ksooo — COMMENTED** [review]
Thanks for picking my comments up.

I would have taken a slightly different approach, though.

I woul have defined a struct `ManagerMessages`containing all the strings, and a pure viertual method that returns a const ref to an instance of this struct. Derived classes override this method, filling the struct (with a static non member variable defined in the cpp file). No maps needed, no map lookups, only reading struct members.

File: xbmc/video/dialogs/GUIDialogVideoManager.h
Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoManager.h:78]: I read conflicting information about the performance of map vs unordered_map. With such a small number of entries, might as well make it an array, using the enum values as index.

**Comment by CrystalP** [comment]
> I would have defined a struct `ManagerMessages`containing all the strings, and a pure viertual method that returns a const ref to an instance of this struct. Derived classes override this method, filling the struct (with a static non member variable defined in the cpp file). No maps needed, no map lookups, only reading struct members.

Updated with something like your suggestion.

The more I think about it, the less I like the idea of AddVideoAssetFilePicker() being a class member. The messages above are not even displayed on the manager dialog itself, but by various popup dialogs.
ChooseVideoAsset() was dumped there for lack of a better location I guess, maybe those 2 should go to a new utility class. That would be a separate PR.

File: xbmc/video/dialogs/GUIDialogVideoManager.h
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManager.h:71]: Please follow our coding guidelines. struct members must start with a small letter.

File: xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp:169]: I don't think you need the `InitializeMessages` method.

Doesn't `static ManagerMessages msg{InitializeMessages{40025, 40039, 40040, 40015, 40034}};` work? Maybe you need to define a respective ctor for `ManagerMessages`.

File: xbmc/video/dialogs/GUIDialogVideoManagerExtras.h
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManagerExtras.h:49]: No need for this. See above.

File: xbmc/video/dialogs/GUIDialogVideoManagerVersions.h
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoManagerVersions.h:121]: No need for this. See above.

**Comment by ksooo** [comment]
> The more I think about it, the less I like the idea of AddVideoAssetFilePicker() being a class member. The messages above are not even displayed on the manager dialog itself, but by various popup dialogs.
ChooseVideoAsset() was dumped there for lack of a better location I guess, maybe those 2 should go to a new utility class. That would be a separate PR.

Hmm. As you introduce `AddVideoAssetFilePicker` with this PR, maybe it would be wise to do one step back and to rewrite the code so that this interim step is not needed at all?

File: xbmc/video/dialogs/GUIDialogVideoManager.h
Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoManager.h]: yes I noticed right after pushing the commit...

File: xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp
Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp]: I though about doing that, but it's not easy to read.

**Comment by CrystalP** [comment]
For now I remove the refactor commits to leave only the original reason for the PR, the removal of unnecessary stream details extraction.

**Comment by CrystalP** [comment]
Jenkins build this please

File: xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp
Comment (ksooo) [reply xbmc/video/dialogs/GUIDialogVideoManagerExtras.cpp]: It’s “standard” with modern c++ I’d say and used widely.

**Comment by ksooo** [comment]
> For now I remove the refactor commits to leave only the original reason for the PR, the removal of unnecessary stream details extraction.

Fair.

**Comment by CrystalP** [comment]
@ksooo would you approve what's left when you have a chance please

**Review by ksooo — APPROVED** [review]
Looks good in general. 

Only concern is that i have formerly seen that code relies on those (sometimes even completely empty) videoinfotags gerated as a side effect of a call to this infamous non-const „GetVideoInfoTag“ method we have. 

So I’m a bit uncertain whether the now  removed call to GetVideoInfoTag causes unexpected side effects elsewhere. 

But this is not a blocker for this PR as i have no idea how to test that before merging. We will find out after merge I guess.

**Comment by CrystalP** [comment]
I thought about it too but in this case, the items being worked on are temporaries in the callers with no relevant processing afterwards.

**Comment by CrystalP** [comment]
Would be good to get away from that non-const GetVideoInfoTag() but it's a large undertaking, different PR if ever.
That would require a transition, maybe starting with a new const GetExistingVideoInfoTag(). Easier to add another better one than renaming all existing to GetOrCreateVideoInfoTag() and disrupt commit history far more.

**Comment by ksooo** [comment]
> Would be good to get away from that non-const GetVideoInfoTag() but it's a large undertaking, different PR if ever.

Yep, i once tried and gave up at some point. And we have the same problem with musicinfotags.

**Pull request merged**
